### PR TITLE
[SPARK-47200][SS] Error class for Foreach batch sink user function error

### PR DESCRIPTION
**Pull request by micheal-o**
### What changes were proposed in this pull request?
Any exception can be thrown from the user provided function for ForEachBatchSink. We want to classify this class of errors. Including errors from Python (Py4JException) and Scala functions.

### Why are the changes needed?
The user provided function can throw any type of error. Using the new error framework for better error messages and classification.

### Does this PR introduce _any_ user-facing change?
Yes, better error message with error class for ForeachBatchSink user function failures.

### How was this patch tested?
Updated existing tests and added a new one. Covers python and Scala.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: docs/sql-error-conditions.md
Comment (MaxGekk) [inline docs/sql-error-conditions.md:1157]: Similar changes made by https://github.com/apache/spark/pull/45226. Could you rebase your PR on the recent master.

Comment (micheal-o) [reply docs/sql-error-conditions.md]: @MaxGekk done

**Comment by micheal-o** [comment]
cc @HeartSaVioR

**Review by HeartSaVioR — APPROVED** [review]
+1
I'll wait for a day (or over the weekend) for @MaxGekk to chime in. I'll merge the PR once he approves or no major comment is made from him.

**Review by MaxGekk — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**
